src/main/java/org/openpnp/machine/reference/vision/ReferenceBottomVision.java

```diff
@@ -99,36 +98,52 @@ private static PartAlignmentOffset findOffsetsPreRotate(Part part, BoardLocation
-            // 0.0765 is chosen so that when adding to half a microstep it is just a bit more than a microstep.
-            // 0.0765 + 0.0567 = 0.1332
-
-            // The check is intended to not introduce further error if the error is small. 
-            // I believe it is to avoid rounding of microsteps, but I'm not clear on that.
-            
-            if (Math.abs(angle) > 0.0765) {
-                angle += 0.0567 * Math.signum(angle);
-            } // rounding
-
-            nozzle.moveTo(new Location(LengthUnit.Millimeters, Double.NaN, Double.NaN, Double.NaN,
-                    placementAngle + angle));
-
-            rect = processPipelineAndGetResult(pipeline, camera, part, nozzle);
-
-            Logger.debug("Result rect {}", rect);
-            Location offsets = VisionUtils.getPixelCenterOffsets(camera, rect.center.x, rect.center.y)
-                                          .derive(null, null, null, Double.NaN);
+        	Location offsets = new Location(location.getUnits());
+        	// Try getting a good fix on the part in multiple passes.
+        	for(int pass = 0;;) {
+	            RotatedRect rect = processPipelineAndGetResult(pipeline, camera, part, nozzle);
+	            camera=(Camera)pipeline.getProperty("camera");
+	            
+	            Logger.debug("Result rect {}", rect);
+	            
+	            // Create the offsets object. This is the physical distance from
+	            // the center of the camera to the located part.
+	            offsets = VisionUtils.getPixelCenterOffsets(camera, rect.center.x, rect.center.y);
+	            
+	            // OpenCV can only tell us the angle of the recognized rectangle in a   
+	            // range of 0° .. 90° as it has no notion of which rectangle side 
+	            // is the true "bottom" side for instance. The angle will abruptly wrap 
+	            // around from 90° to 0° (and vice versa), as another of the four sides 
+	            // becomes the "bottom" side, as perceived by OpenCV.    
+	            // We can assume that the part is never picked more than +/-45º rotated.
+	            // So we change the wrapping-around into one that produces a range from 
+	            // -45° .. +45°. See angleNorm():
+	            angle = angleNorm(location.getRotation() + rect.angle);
+	            
+	            // When we rotate the nozzle later to compensate the angle, the X, Y offsets 
+	            // will change too, as the off-center part rotates around the nozzle axis.
+	            // So we need to compensate for that.
+	            // TODO: verify angle sign here. 
+	            offsets = offsets.rotateXy(-angle)
+	            			.derive(null, null,	null, -angle);
+	                
+	            if (++pass > 2) {
+	            	// Maximum number of passes. 
+	            	break;
+	            }
+	            
+	            if (Math.sqrt(offsets.getX()*offsets.getX() + offsets.getY()*offsets.getY()) < 2.0
```

Comment:
Is the `2` here in millimeters? Or is it unitless and I don't quite understand? If millimeters, this should be corrected to check against the correct units. Additionally, 2mm or 10 degrees of offset seems like quite a lot. Should we make this much smaller, or perhaps configurable?

Reply:
This is still TODO. See the reply to [your question](https://github.com/openpnp/openpnp/pull/815/files/6ec8856de8f50e810c26fa95ed1547e34238aea4#issuecomment-483909834).